crates/chia-sdk-driver/src/layers/p2_eip712_message_layer.rs

```diff
@@ -0,0 +1,298 @@
+    }
+
+    pub fn hash_to_sign(&self, coin_id: Bytes32, delegated_puzzle_hash: Bytes32) -> Bytes32 {
+        /*
+        bytes32 messageHash = keccak256(abi.encode(
+            typeHash,
+            coin_id,
+            delegated_puzzle_hash
+        ));
+        */
+        let mut to_hash = Vec::new();
+        to_hash.extend_from_slice(&P2Eip712MessageLayer::type_hash());
+        to_hash.extend_from_slice(&coin_id);
+        to_hash.extend_from_slice(&delegated_puzzle_hash);
+
+        let message_hash = keccak256(&to_hash);
+
+        let mut to_hash = Vec::new();
+        to_hash.extend_from_slice(&self.prefix_and_domain_separator());
+        to_hash.extend_from_slice(&message_hash);
+
+        keccak256(&to_hash).into()
+    }
+}
+
+impl Layer for P2Eip712MessageLayer {
+    type Solution = P2Eip712MessageSolution<NodePtr, NodePtr>;
+
+    fn construct_puzzle(&self, ctx: &mut SpendContext) -> Result<NodePtr, DriverError> {
+        let curried = CurriedProgram {
+            program: ctx.p2_eip712_message_puzzle()?,
+            args: P2Eip712MessageArgs {
+                prefix_and_domain_separator: self.prefix_and_domain_separator().to_vec().into(),
+                type_hash: P2Eip712MessageLayer::type_hash(),
+                pubkey: self.pubkey.to_vec().into(),
+            },
+        };
+        ctx.alloc(&curried)
+    }
+
+    fn construct_solution(
+        &self,
+        ctx: &mut SpendContext,
+        solution: Self::Solution,
+    ) -> Result<NodePtr, DriverError> {
+        ctx.alloc(&solution)
+    }
+
+    fn parse_puzzle(_allocator: &Allocator, _puzzle: Puzzle) -> Result<Option<Self>, DriverError> {
+        Ok(None)
+    }
+
+    fn parse_solution(
+        allocator: &Allocator,
+        solution: NodePtr,
+    ) -> Result<Self::Solution, DriverError> {
+        Ok(P2Eip712MessageSolution::from_clvm(allocator, solution)?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::assert_puzzle_hash;
+
+    use super::*;
+    use chia_consensus::consensus_constants::TEST_CONSTANTS;
+    use chia_protocol::Bytes;
+    use chia_sdk_test::Simulator;
+    use chia_sdk_types::Conditions;
+    use clvm_traits::clvm_quote;
+    use clvm_utils::ToTreeHash;
+    use clvmr::chia_dialect::ENABLE_KECCAK_OPS_OUTSIDE_GUARD;
+    use clvmr::reduction::Reduction;
+    use clvmr::serde::node_from_bytes;
+    use ethers::core::rand::thread_rng;
+    use ethers::prelude::*;
+    use ethers::signers::LocalWallet;
+    use k256::ecdsa::signature::hazmat::PrehashVerifier;
+    use k256::ecdsa::{Signature as K1Signature, SigningKey, VerifyingKey as K1VerifyingKey};
+    use rstest::rstest;
+
+    #[test]
+    fn test_puzzle_hash() -> anyhow::Result<()> {
+        assert_puzzle_hash!(P2_EIP712_MESSAGE_PUZZLE => P2_EIP712_MESSAGE_PUZZLE_HASH);
+        Ok(())
+    }
+
+    #[test]
+    fn test_softfork_cost() -> anyhow::Result<()> {
+        let ctx = &mut SpendContext::new();
+        let puzzle_bytes =
```

Comment:
What puzzle is this?

Reply:
code running inside softfork; will add comment